ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,559 @@
+	minCache = 16
+
+	// minHandles is the minimum number of files handles to allocate to the open
+	// database files.
+	minHandles = 16
+
+	// metricsGatheringInterval specifies the interval to retrieve leveldb database
+	// compaction, io and pause stats to report to the user.
+	metricsGatheringInterval = 3 * time.Second
+)
+
+// Database is a persistent key-value store based on the pebble storage engine.
+// Apart from basic data storage functionality it also supports batch writes and
+// iterating over the keyspace in binary-alphabetical order.
+type Database struct {
+	fn string     // filename for reporting
+	db *pebble.DB // Underlying pebble storage engine
+
+	compTimeMeter       metrics.Meter // Meter for measuring the total time spent in database compaction
+	compReadMeter       metrics.Meter // Meter for measuring the data read during compaction
+	compWriteMeter      metrics.Meter // Meter for measuring the data written during compaction
+	writeDelayNMeter    metrics.Meter // Meter for measuring the write delay number due to database compaction
+	writeDelayMeter     metrics.Meter // Meter for measuring the write delay duration due to database compaction
+	diskSizeGauge       metrics.Gauge // Gauge for tracking the size of all the levels in the database
+	diskReadMeter       metrics.Meter // Meter for measuring the effective amount of data read
+	diskWriteMeter      metrics.Meter // Meter for measuring the effective amount of data written
+	memCompGauge        metrics.Gauge // Gauge for tracking the number of memory compaction
+	level0CompGauge     metrics.Gauge // Gauge for tracking the number of table compaction in level0
+	nonlevel0CompGauge  metrics.Gauge // Gauge for tracking the number of table compaction in non0 level
+	seekCompGauge       metrics.Gauge // Gauge for tracking the number of table compaction caused by read opt
+	manualMemAllocGauge metrics.Gauge // Gauge for tracking amount of non-managed memory currently allocated
+
+	quitLock sync.Mutex      // Mutex protecting the quit channel access
+	quitChan chan chan error // Quit channel to stop the metrics collection before closing the database
+
+	log log.Logger // Contextual logger tracking the database path
+
+	activeComp          int       // current number of active compactions
+	compStartTime       time.Time // the start time of the earliest currently-active compaction
+	compTime            int64     // total time spent in compaction in ns
+	seekCompCount       int64     // total number of compactions caused by reads
+	level0Comp          uint32    // total number of level-zero compactions
+	nonLevel0Comp       uint32    // total number of non level-zero compactions
+	writeDelayStartTime time.Time // the start time of the latest write stall
+	writeDelayCount     int64     // total number of write stall counts
+	writeDelayTime      int64     // total time spent in write stalls
+}
+
+func (d *Database) OnCompactionBegin(info pebble.CompactionInfo) {
```

Comment:
All these `On..` - handlers: looks to me like they can be un-exported (if not, you should add a docstring to them.)